### PR TITLE
multi-electrode adaptation

### PR DESCRIPTION
**Pull request by TheLemonPig**
This update has:

-added condition for the number of electrodes per trial. The default is n_channels=1, which is backwards compatible with single electrode training
-removal of stale comments from trainer and gan_training_main.py. This is mostly to make the code more readable.
-dataloader.py includes a few lines to reshape the data by the number of electrodes
-trainer.py contains a few lines to move electrodes from the 2nd to the 1st dimension
-trainer.py contains a few minimal changes to code to accommodate the extra dimension

Still needs testing:

   -DDP Trainer with multiple-electrodes
   -confirming that all functionality still works e.g. plotting directly from checkpoint, and plotting PCA
   **-single electrode recovery is still successful**
   **-multi-electrode recovery is successful**
   ...

### CODE REVIEW

**Comment by chadcwilliams** [comment]
I'm getting some issues with this branch. I have so far only worked on the gan_training_main.py file but it's functionality seems to be changed in an unexpected way. Specifically, it now always ignores all parameters given and tries to run with defaults. 

For example, `python gan_training_main.py help` should output parameter information, but instead tries to run the GAN with the default data `data/ganAverageERP_mini.csv`. I get the same outcome when running `python gan_training_main.py path_dataset=data/ganTrialElectrodeERP.csv n_epochs=8000`. Additionally, the code should identify any parameters given to it that are not real parameters and output a warning saying so, but  `python gan_training_main.py fake_parameter` again tries to run with defaults.

File: gan_training_main.py [hunk 1]
```diff
@@ -22,12 +22,11 @@
       - the remaining columns contain the time-series data
   - set the configuration parameters (Training configuration; Data configuration; GAN configuration)"""
 
-# TODO: update the training process and the GANs so the distinction between TTS-GAN and other GANs is not necessary
 
 if __name__ == '__main__':
     """Main function of the training process."""
 
-    # sys.argv = ["path_dataset=data/ganAverageERP_len100.csv", "patch_size=20", "conditions=Condition"]
+    sys.argv = ["path_dataset=data/ganAverageERP_mini.csv", "patch_size=20", "conditions=Condition", 'n_channels=2', 'n_epochs=25']
```
Comment (chadcwilliams) [inline gan_training_main.py:29]: Check my overall comment on the PR but this line is causing the issue. By uncommenting this line, you override any parameters inputted by the user.

Comment (TheLemonPig) [reply gan_training_main.py:29]: Ah I see! Sorry about that. I will make sure to leave that commented out in commits then

Comment (TheLemonPig) [inline gan_training_main.py:29]:
```suggestion
    # sys.argv = ["path_dataset=data/ganAverageERP_mini.csv", "patch_size=20", "conditions=Condition", 'n_channels=2', 'n_epochs=25']
```

[the remark this answers is not in the document]
Comment (TheLemonPig) [reply gan_training_main.py]: All good these sound like very worthwhile changes to the code. The first change will be very easy, so I can go ahead with that. For the second change, where the user is able to give a list of the electrodes to use, I am unsure how to determine which channels those names correspond to. Currently, the data labels electrodes by number. Is there a conventional mapping from electrode number to electrode name? Or will the data usually label the electrodes by name?

[the remark this answers is not in the document]
Comment (chadcwilliams) [reply gan_training_main.py]: Thanks Joshua! If the first change is trivial, it might be worth doing then. I know it's not much of a problem for users to have to manually define the number of electrodes, but it is still an extra step. For example, I tried to run this and had to open my training file to remember how many electrodes the data had *facepalm*. 

I would say don't do the latter suggestion right now - it's not necessary and just functionality we can add in the future (I just wanted to mention it). But to answer your question - no, there is no standard link between the index numbers and electrode names. In this example, I would instead give you a csv file where the electrodes were names instead of numbers.

**Comment by TheLemonPig** [comment]
ddp_training.py doesn't work for me for a different reason that I think is hardware related. I made the changes that I had made in the past to make it work. I would be curious to know if it works now.

**Review by chadcwilliams — CHANGES_REQUESTED** [review]
I have tested these changes quite thoroughly over the last week and a half. I have tested on 50 participants with 3 to 64 electrodes and from 2000 to 20000 training epochs. Unfortunately, none of the GANs converged and there was some odd convergence patterns that I have not seen before with the original EEG-GAN (I have kept Joshua updated on all this throughout the process).

As such, I believe there needs to be a thorough code review (by Joshua) to ensure everything is in order and more testing.

**Comment by TheLemonPig** [comment]
> I have tested these changes quite thoroughly over the last week and a half. I have tested on 50 participants with 3 to 64 electrodes and from 2000 to 20000 training epochs. Unfortunately, none of the GANs converged and there was some odd convergence patterns that I have not seen before with the original EEG-GAN (I have kept Joshua updated on all this throughout the process).
> 
> As such, I believe there needs to be a thorough code review (by Joshua) to ensure everything is in order and more testing.

I have run further tests and can confirm most things I can think of are in order. I will continue to consider more tests. I have tried running on adjust meta-parameters based on plot losses chad received on training to see if this will help with convergence. So far this looks like it is working. I will get back as soon as I have a plot of average generated data that captures the condition pattern

**Review by chadcwilliams — APPROVED** [review]
Tests have shown it to be working now. Still struggles with EEG data, but that's likely do to noise in the data. Worked with synthetic data.
